File: Backend/Services/UnitService.cs

using System.Data;
using Backend.Dtos;
using Backend.Models;
using Backend.Repositories.Interfaces;
using Backend.Services.Interfaces;

namespace Backend.Services;

public class UnitService(IUnitRepository unitRepository) : IUnitService
{
    private readonly IUnitRepository _unitRepository = unitRepository;

    public async Task<IEnumerable<GetUnitDto>> GetAllUnits(Guid userId)
    {
        var units = await _unitRepository.GetAllUnits(userId);

        var unitDtos = units.Select(u => new GetUnitDto
        {
            Id = u.Id,
            Title = u.Title,
            CreatorId = u.CreatorId
        });

        return unitDtos;
    }

    public async Task EditUnit(Guid unitId, NewEditUnitDto unitEdited)
    {
        var unit = await _unitRepository.GetSingleUnit(unitId);

        if (unit == null)
        {
            throw new KeyNotFoundException($"Unit with id {unitId} does not exist.");
        }

        if (unit.CreatorId == null)
        {
            throw new ConstraintException($"Cannot modify pre-defined units.");
        }

        var nameAlreadyTaken = await _unitRepository.CheckUnitExistsByName(unit.CreatorId, unitEdited.Title, unitId);

        if (nameAlreadyTaken)
        {
            throw new InvalidOperationException($"Unit {unitEdited.Title} already exists.");
        }

        await _unitRepository.EditUnit(unit, unitEdited);

    }

    public async Task<GetUnitDto> AddNewUnit(Guid creatorId, NewEditUnitDto unitNew)
    {
        var alreadyExists = await _unitRepository.CheckUnitExistsByName(creatorId, unitNew.Title);

        if (alreadyExists)
        {
            throw new InvalidOperationException($"Unit {unitNew.Title} already exists.");
        }

        var unit = new Unit
        {
            Id = Guid.NewGuid(),
            Title = unitNew.Title,
            CreatorId = creatorId
        };

        await _unitRepository.AddNewUnit(unit);

        var unitDto = new GetUnitDto
        {
            Id = unit.Id,
            Title = unit.Title,
            CreatorId = unit.CreatorId
        };

        return unitDto;
    }

    public async Task DeleteUserDefinedUnit(Guid unitId)
    {
        var unit = await _unitRepository.GetSingleUnit(unitId);

        if (unit == null)
        {
            throw new KeyNotFoundException($"Unit with id {unitId} does not exist.");
        }

        if (unit.CreatorId == null)
        {
            throw new ConstraintException("Cannot delete pre-defined units.");
        }

        await _unitRepository.DeleteUnit(unit);
    }

    public async Task BatchDeleteUnits(BatchDeleteDto idsToDelete)
    {
        var unitsToDelete = await _unitRepository.GetMultipleUnits(idsToDelete.Ids);

        var missingIds = idsToDelete.Ids.Except(unitsToDelete.Select(unit => unit.Id));

        if (missingIds.Any())
        {
            throw new KeyNotFoundException($"Units with ids {missingIds} do not exist.");
        }

        if (unitsToDelete.Any(unit => unit.CreatorId == null))
        {
            throw new ConstraintException("Cannot delete pre-defined units.");
        }

        await _unitRepository.DeleteBatchUnits(unitsToDelete);
    }
}

File: Frontend/src/api/services/unitService.ts
import type { BatchDeleteDto } from "../../domain/models/batchDeleteDto";
import type { GetUnitDto } from "../../domain/models/getUnitDto";
import type { NewEditUnitDto } from "../../domain/models/newEditUnitDto";
import BaseService from "./baseService";

class UnitService extends BaseService {
  constructor() {
    super("api/unit");
  }

  public async getAllUnits(): Promise<Array<GetUnitDto>> {
    let allUnits = await this.get<Array<GetUnitDto>>("all");

    return allUnits;
  }

  public async createUnit(newUnit: NewEditUnitDto): Promise<GetUnitDto> {
    let addedUnit = await this.post<NewEditUnitDto, GetUnitDto>("", newUnit);

    return addedUnit;
  }

  public async editUnit(
    unitId: string,
    editedUnit: NewEditUnitDto
  ): Promise<void> {
    await this.put<NewEditUnitDto, void>(`${unitId}`, editedUnit);
  }

  public async deleteUnit(unitId: string): Promise<void> {
    await this.delete<void>(`${unitId}`);
  }

  public async deleteBatchUnits(unitsToDelete: BatchDeleteDto): Promise<void> {
    await this.post<BatchDeleteDto, void>("delete-multiple", unitsToDelete);
  }
}

const unitService = new UnitService();

export default unitService;
